File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var searchRange = function(nums, target) {
    let start = 0;
    let end = nums.length-1;
    let result = [-1, -1];

    while(start <= end){ 
        let mid = Math.floor((start + end) / 2);
        
        if(nums[mid] === target){
            result[0] = mid; 
            result[1] = mid; 
            
            let left = mid - 1;
            while (left >= 0 && nums[left] === target) {
                result[0] = left;
                left--;
            }
            
            let right = mid + 1;
            while (right < nums.length && nums[right] === target) {
                result[1] = right;
                right++;
            }
            
            return result;
        } else if(nums[mid] < target){
            start = mid + 1;
        } else {
            end = mid - 1;
        }
    }
    
    return result; 
};

File: 0034-find-first-and-last-position-of-element-in-sorted-array/0034-find-first-and-last-position-of-element-in-sorted-array.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        int left = 0;
        int right = n - 1;
        vector<int> ans = {-1, -1};

        while (left <= right) {
            int mid = (left + right) / 2;

            if (nums[mid] == target) {
                ans[0] = mid;
                ans[1] = mid;

                int low = mid - 1;
                while (low >= 0 && nums[low] == target) {
                    ans[0] = low;
                    low--;
                }

                int high = mid + 1;
                while (high < n && nums[high] == target) {
                    ans[1] = high;
                    high++;
                }

                break; 
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return ans;
    }
};
